lib/Sema/CSGen.cpp

```diff
@@ -1539,23 +1551,51 @@ namespace {
             CS.getConstraintLocator(expr),
             {expr->getArgumentLabels(),
              expr->getUnlabeledTrailingClosureIndex()});
-        return baseTy;
+        return outputTy;
       }
-
-      // Otherwise, the member needs to be convertible to the base type.
-      CS.addConstraint(ConstraintKind::Conversion, memberTy, baseTy,
-        CS.getConstraintLocator(expr, ConstraintLocator::RValueAdjustment));
       
-      // The member type also needs to be convertible to the context type, which
-      // preserves lvalue-ness.
-      auto resultTy = CS.createTypeVariable(memberLocator,
-                                            TVO_CanBindToLValue |
-                                            TVO_CanBindToNoEscape);
-      CS.addConstraint(ConstraintKind::Conversion, memberTy, resultTy,
-                       memberLocator);
-      CS.addConstraint(ConstraintKind::Equal, resultTy, baseTy,
-                       memberLocator);
-      return resultTy;
+      // Otherwise, the result is just the type of the member.
+      return memberTy;
+    }
+
+    Type visitUnresolvedMemberChainResultExpr(
+        UnresolvedMemberChainResultExpr *expr) {
+      auto *tail = expr->getSubExpr();
+      assert(isa<UnresolvedMemberExpr>(tail) || isa<UnresolvedDotExpr>(tail) ||
+             isa<CallExpr>(tail) || isa<BindOptionalExpr>(tail) ||
+             isa<ForceValueExpr>(tail) ||
+             isa<SubscriptExpr>(tail) &&
+                 "Unexpected expression at end of unresolved member chain");
+
+      auto memberTy = CS.getType(tail);
+      auto *base = TypeChecker::getUnresolvedMemberChainBase(tail);
+
+      // FIXME: This is a workaround for SR-13357, should not be necessary.
+      // Copy any type variable options from the result of the tail member to
+      // the result of the entire chain.
+      unsigned additionalOptions = 0;
+      if (auto *tvt = memberTy->getAs<TypeVariableType>())
+        additionalOptions = tvt->getImpl().getRawOptions();
+
+      // The contextual type (represented with a new type variable) must equal
+      // the base type.
+      auto locator = CS.getConstraintLocator(
+          expr, ConstraintLocator::UnresolvedMemberChainResult);
+      auto tvo = additionalOptions | TVO_CanBindToHole | TVO_CanBindToNoEscape;
+      auto chainResultTy = CS.createTypeVariable(locator, tvo);
+      auto chainBaseTy = UnresolvedBaseTypes.find(base)->second;
```

Comment:
Should this use `getUnresolvedBaseType`?

Reply:
Yes, thank you!